File: commands.experiments/fs-access.pl
#!/usr/bin/env perl

# Executed in   22.70 millis    fish           external
#    usr time    8.94 millis    0.12 millis    8.82 millis
#    sys time   11.10 millis    1.12 millis    9.98 millis

# Requirements
use strict;
use warnings;

# Check for invalid arguments
if ( @ARGV != 1 || !$ARGV[0] ) {
	exit 22;  # EINVAL 22 Invalid argument
}

# Check for existence and accessibility
# If success, it exists and is accessible
# If failure, check if non-existence was due to non-accessible, or actually non-existence
if (stat($ARGV[0])) {
    exit 17;  # EEXIST 17 File exists
} elsif ($! == 2) {  # ENOENT
    exit 2;   # No such file or directory
} elsif ($! == 13) {  # EACCES
    exit 13;  # Permission denied
} else {
	exit $!; # Return the error code
}

File: commands.experiments/fs-access.py
#!/usr/bin/env python3
# trunk-ignore-all(flake8/E501)

# Executed in   51.94 millis    fish           external
#    usr time   30.49 millis    0.11 millis   30.38 millis
#    sys time   13.91 millis    1.10 millis   12.82 millis

# Requirements
import os
import sys

# Check for invalid arguments
if len(sys.argv) != 2 or not sys.argv[1]:
    sys.exit(22)  # EINVAL 22 Invalid argument

# Check for existence and accessibility
# If success, it exists and is accessible
# If failure, check if non-existence was due to non-accessible, or actually non-existence
try:
    os.stat(sys.argv[1])
    sys.exit(17)  # EEXIST 17 File exists
except FileNotFoundError:
    sys.exit(2)  # ENOENT 2 No such file or directory
except PermissionError:
    sys.exit(13)  # EACCES 13 Permission denied
